File: backend/core/ip_profile.py
class IPProfile:
    """
    IPProfile — Tracks the behavior of a single IP address.
    """

    def __init__(self, ip: str):
        self.ip = ip
        self.login_fail = 0
        self.login_success = 0
        self.request = 0
        self.risk_score = 0

    def increase_fail(self):
        self.login_fail += 1

    def increase_success(self):
        self.login_success += 1

    def increase_request(self):
        self.request += 1

    def calculate_risk_score(self):
        self.risk_score = (self.login_fail * 3) + (self.request * 1) - (self.login_success * 2)
        if self.risk_score < 0:
            self.risk_score = 0

    def get_risk_score(self) -> int:
        return self.risk_score

    def get_ip(self) -> str:
        return self.ip

    def get_login_fail(self) -> int:
        return self.login_fail

    def get_login_success(self) -> int:
        return self.login_success

    def get_request(self) -> int:
        return self.request

File: archive/java_src/core/IPProfile.java
package core;

/**
 * IPProfile — Tracks the behavior of a single IP address.
 *
 * Counters: LOGIN_FAIL, LOGIN_SUCCESS, REQUEST
 * Risk score formula: (loginFail * 3) + (request * 1) - (loginSuccess * 2)
 * Score is clamped to a minimum of 0.
 *
 * Note: This class is currently NOT used in the main flow.
 * Equivalent logic is embedded directly in SecurityBot.
 * Kept here for future refactoring (decoupling scoring from SecurityBot).
 */
public class IPProfile {

    /** IP address this profile belongs to */
    private String ip;

    /** Number of failed login attempts */
    private int loginFail;

    /** Number of successful logins */
    private int loginSuccess;

    /** Number of requests sent */
    private int request;

    /** Computed risk score */
    private int riskScore;

    /**
     * Creates a new IPProfile for the given IP address.
     * All counters start at 0.
     *
     * @param ip IP address (e.g. "192.168.1.15")
     */
    public IPProfile(String ip) {
        this.ip = ip;
    }

    /** Increments the LOGIN_FAIL counter by 1. */
    public void increaseFail() {
        loginFail++;
    }

    /** Increments the LOGIN_SUCCESS counter by 1. */
    public void increaseSuccess() {
        loginSuccess++;
    }

    /** Increments the REQUEST counter by 1. */
    public void increaseRequest() {
        request++;
    }

    /**
     * Calculates the risk score using the weighted formula:
     *   riskScore = (loginFail * 3) + (request * 1) - (loginSuccess * 2)
     *
     *   - LOGIN_FAIL  weight x3 — clearest indicator of attack
     *   - REQUEST     weight x1 — could be normal traffic
     *   - LOGIN_SUCCESS weight -2 — reduces suspicion
     *
     * Result is clamped to >= 0.
     */
    public void calculateRiskScore() {
        riskScore = (loginFail * 3) + (request * 1) - (loginSuccess * 2);
        if (riskScore < 0) {
            riskScore = 0;
        }
    }

    // ===================== GETTERS =====================

    /** @return Computed risk score (call calculateRiskScore() first) */
    public int getRiskScore() {
        return riskScore;
    }

    /** @return IP address */
    public String getIp() {
        return ip;
    }

    /** @return Number of LOGIN_FAIL events */
    public int getLoginFail() {
        return loginFail;
    }

    /** @return Number of LOGIN_SUCCESS events */
    public int getLoginSuccess() {
        return loginSuccess;
    }

    /** @return Number of REQUEST events */
    public int getRequest() {
        return request;
    }
}